### PR TITLE
Optimize degree() to avoid unnecessary expansion of factored polynomials

### PR DESCRIPTION
**Pull request by sohaa-khan11**
Fixes degree issues of #6322

The degree() function previously relied on poly_from_expr, which eagerly expanded expressions by default. This caused severe performance issues for factored polynomials such as (x + 1)**10000.
This PR adds a conservative structural fast-path to degree() that computes the degree recursively for safe cases (Pow, Mul, and Add with a unique maximal-degree term) without expanding the expression.
If cancellation is possible or the structure is unsafe, the implementation falls back to the existing logic unchanged, preserving correctness.

The optimization is intentionally limited to degree() and does not modify poly_from_expr defaults or global expansion behavior.
Regression tests are included to validate both fast-path behavior and fallback cases.

review suggestions were provided by an AI tool

* polys
  * Improved performance of `degree()` for factored polynomials by avoiding unnecessary expansion in certain cases.

### CODE REVIEW

**Comment by RushabhMehta2005** [comment]
I don't know what this PR is doing but simply asking an LLM to optimise something and then making a PR out of it is unacceptable

**Comment by sohaa-khan11** [comment]
To clarify what I did here: I didn’t just ask an LLM to “optimize” something and submit the result. This issue (#6322) is a performance problem that I investigated manually by tracing degree() through poly_from_expr and identifying where expansion was happening.
AI was used only as a discussion/review aid; the design, implementation, and tests were done manually. I’m happy to revise the approach or explain any part of the change.

**Comment by RushabhMehta2005** [comment]
Could you run this through a profiler? i.e before the fix and after the fix, is there any actual impact on performance?

**Comment by ayush-bothra** [comment]
Ive just checked for the first assert in the tests

```text
In [1]: from sympy import symbols, degree

In [2]: x = symbols('x')

In [3]: %time degree((x+1)**10000, x)
CPU times: user 8.18 s, sys: 125 ms, total: 8.3 s
Wall time: 8.5 s
Out[3]: 10000
```

currently degree takes about 8.5s for this code

if the new addition is faster, and by how much, would be a good thing to add in the other comments section so that its easier to check the performance gain

**Comment by sohaa-khan11** [comment]
image
I ran a simple before/after timing comparison on my machine.
On upstream/master, calling
degree((x + 1)**10000, x)
took about 27 seconds, due to full expansion triggered by poly_from_expr.
With this PR applied, the same call completes in about 7.7e-05 seconds

**Comment by RushabhMehta2005** [comment]
For failing authors test, follow [this](https://docs.sympy.org/latest/contributing/new-contributors-guide/workflow-process.html#mailmap-instructions) carefully step by step.

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -4938,6 +4989,12 @@ def degree(f, gen=0):
     else:
         isNum = f.is_Number
         if not isNum:
+            # Conservative fast-path to avoid unnecessary expansion
+            if not gen_is_Num:
+                fast_deg = _degree_fast(f, gen)
+                if fast_deg is not None:
+                    return Integer(fast_deg) if isinstance(fast_deg, int) else fast_deg
```
Comment (smichr) [inline sympy/polys/polytools.py]:
```suggestion
                    return fast_deg
```

File: sympy/polys/polytools.py [hunk 2]
```diff
@@ -4903,6 +4903,57 @@ def _update_args(args, key, value):
...
+def _degree_fast(expr, gen):
+    if expr.is_Number:
+        if expr is S.Zero:
+            return S.NegativeInfinity
+        return 0
+
+    if expr == gen:
+        return 1
+    if expr.is_Symbol:
+        return 0
```
Comment (smichr) [inline sympy/polys/polytools.py]:
```suggestion
        return S.Zero

    if expr == gen:
        return S.One
    if expr.is_Symbol:
        return S.Zero
```

File: sympy/polys/tests/test_polytools.py [hunk 3]
```diff
@@ -4264,3 +4264,20 @@ def test_schur_conditions():
 
     assert any(c <= 0 for c in p6.set_domain(QQ).schur_conditions())
     assert any(c <= 0 for c in p6.set_domain(EXRAW).schur_conditions())
+
+
+def test_degree_optimization():
```
Comment (smichr) [inline sympy/polys/tests/test_polytools.py]: Maybe something like `degree((2*exp(x)**4+1)**6 + 1, exp(x))==24` would be good to add.

**Comment by smichr** [comment]
The recursive simplicity is so nice.

Does `as_leading_term` benefit from this?

**Comment by sohaa-khan11** [comment]
Thanks! as_leading_term already works structurally and avoids poly_from_expr, so it doesn’t benefit directly from this change. This patch applies a similar structural approach to degree() to avoid unnecessary expansion.

**Comment by ayush-bothra** [comment]
> I ran a simple before/after timing comparison on my machine. On upstream/master, calling degree((x + 1)**10000, x) took about 27 seconds, due to full expansion triggered by poly_from_expr. With this PR applied, the same call completes in about 7.7e-05 seconds

thats great, for simple expressions this will be a massive speedup, the issue linked also requested work on `LC`
maybe u can look into that as well

**Comment by sohaa-khan11** [comment]
> > I ran a simple before/after timing comparison on my machine. On upstream/master, calling degree((x + 1)**10000, x) took about 27 seconds, due to full expansion triggered by poly_from_expr. With this PR applied, the same call completes in about 7.7e-05 seconds
> 
> thats great, for simple expressions this will be a massive speedup, the issue linked also requested work on `LC` maybe u can look into that as well

Thanks! I’d be happy to look into LC separately in a follow-up.

**Comment by smichr** [comment]
> issue linked also requested work on LC

This can be treated as a separate issue. It is easy for a PR to grow considerably with each new phase of the problem that is tackled. This is a nice sized change already.

If you decide to hold at this PR then change "FIXES #6322" to "FIXES DEGREE ISSUE OF #6322" -- insertion of words between the action word (fix(es), close(s)) will not close the issue. Then on the issue page you can just say that the degree issue was addressed but the remaining work is __.

**Comment by sohaa-khan11** [comment]
> > issue linked also requested work on LC
> 
> This can be treated as a separate issue. It is easy for a PR to grow considerably with each new phase of the problem that is tackled. This is a nice sized change already.
> 
> If you decide to hold at this PR then change "FIXES #6322" to "FIXES DEGREE ISSUE OF #6322" -- insertion of words between the action word (fix(es), close(s)) will not close the issue. Then on the issue page you can just say that the degree issue was addressed but the remaining work is __.

I’ll keep this PR focused on degree() and update the wording so it doesn’t auto-close the issue. I’ll leave a note on #6322 describing the remaining work.

File: sympy/polys/tests/test_polytools.py (hunk 3 above)
Comment (sohaa-khan11) [reply sympy/polys/tests/test_polytools.py]: I’ve added a test along those lines:
degree((2*exp(x)**4 + 1)**6 + 1, exp(x)) == 24, and verified that it passes. Thanks for the suggestion!

**Comment by oscarbenjamin** [comment]
This is AI slop that fixes nothing.

**Comment by oscarbenjamin** [comment]
> **AI Generation Disclosure**
> review suggestions were provided by an AI tool

Total BS. All the code is AI. Non-disclosure is a clear breach of policy:
https://docs.sympy.org/dev/contributing/ai-generated-code-policy.html

**Comment by RushabhMehta2005** [comment]
Thanks for closing this.

**Comment by smichr** [comment]
Re-opening to allow technical discussion. This PR has received substantive human review and does not appear to violate the current AI policy. If there are technical reasons the change is unacceptable (correctness, maintainability, performance, or scope), please raise them here.

File: sympy/polys/polytools.py [hunk 4]
```diff
@@ -4903,6 +4903,57 @@ def _update_args(args, key, value):
     return args
 
 
+@public
+def _degree_fast(expr, gen):
+    if expr.is_Number:
+        if expr is S.Zero:
```
Comment (smichr) [inline sympy/polys/polytools.py]:
```suggestion
        if expr.is_zero:  # 0.0 or 0
```

**Comment by ayush-bothra** [comment]
The pr works for very simple degree calculations, and doesn't simplify cancellations
As I had mentioned on the original issue
@smichr s idea of hiding generators other than required can possibly be paired up with ring calculations as those will be faster than current math

@oscarbenjamin I can test this idea and put my results here if it is a viable plan of action

**Comment by oscarbenjamin** [comment]
> > **AI Generation Disclosure**
> > review suggestions were provided by an AI tool
> 
> Total BS. All the code is AI. Non-disclosure is a clear breach of policy: https://docs.sympy.org/dev/contributing/ai-generated-code-policy.html

Sorry I should not have responded like this. That was rude and unprofessional.

Let me explain more clearly what I meant which is that I can tell just by looking in the PR that the code is AI generated and I don't think that "review suggestions were provided by an AI tool" is an accurate reflection of how AI was used.

Unfortunately sympy is overwhelmed by large numbers of new contributors opening PRs many of which are low quality and generated by AI and so we have a policy that prohibits the *undisclosed* use of AI to write the code. I closed the PR above because I think that this is all written by AI and that was not adequately disclosed. That may have been a misunderstanding on your part though that "review suggestions" perhaps reads differently to you rather than me.

There are a number of problems with AI generated code but one of them is that it is just very easy to just add more and more code and then end up with the codebase being an unmaintainable mess. I think it is possible to use AI to generate good code but it requires very strong controls and more effort on design rather than falling into a pattern of just more and more code. The same problems can happen without using AI but are just less likely because when human effort is involved in writing the code people tend to prioritise having less code naturally and the process of actually writing the code causes them to think more about the overall design.

The proper fix for this issue is that there should be a version of Poly that does not expand everything rather than having specific functions like this just for degree and LC. Otherwise it is better not to add ad-hoc code for particular cases because it is increasing the surface area for bugs while not really delivering much in the way of actual improvement of sympy's features.

File: sympy/polys/polytools.py
Comment (smichr) [inline sympy/polys/polytools.py]:
```suggestion
    gen = sympify(gen, strict=True)
    gen_is_Num = gen.is_Number
```

Comment (smichr) [inline sympy/polys/polytools.py]:
```suggestion
```

**Comment by smichr** [comment]
There isn’t an unevaluated Poly today, and creating a fully lazy polynomial type is a large project in itself.

But degree/LC/LM are queries that can often be answered from the expression tree without expansion.

I don't view a conservative fast-path as "ad hoc"; it’s an optimization of an existing method that can work with expr or Poly (as opposed to Poly.degree), with fallback to the canonical polynomial conversion when needed.

**Comment by smichr** [comment]
This is what I have worked on so far:

[deleted in favor of getting other things fixed/decided first]

**Comment by sohaa-khan11** [comment]
Thanks for reopening and for pushing this forward. I apologize for any confusion around the earlier AI disclosure — that wasn’t my intent. I like the conservative fast-path approach and the clear fallback to Poly.
I can add tests, check edge cases, and do basic benchmarking if that’s useful.

**Comment by smichr** [comment]
I update [comment](https://github.com/sympy/sympy/pull/29051#issuecomment-3834801638). Unless degree is deprecated to only allow for expressions and symbol generators, I don't know how to make it much simpler. So I am open to suggestions!

**Comment by smichr** [comment]
Why don't we hold off on changes here until #29090 can be addressed.

**Comment by sohaa-khan11** [comment]
That makes sense to me. A lot of the complexity seems to come from unclear generator selection for non-polynomial expressions, and fixing that first would make later changes to degree much easier and safer.

**Comment by smichr** [comment]
What about using Poly(..., expand=False).gens to act as a gate for the check for whether the function is polynomial or not? I am just leaving this as a breadcrumb. Maybe it would be useful in `is_polynomial` or maybe it would be useful for determining whether there is only one generator in an expression. That is something that I was struggling with because creating an *expanded* poly is an expensive way to find the generators. But Poly can still find generators without expanding. *What* it finds will be different than when the expression is expanded, but I think it could help a lot. I did some work on trying to identify generators in #22267 but I think using Poly directly would be better.

```python
>>> Poly((1+x)**4+(1+x)**2-x-3,expand=False).gens
(x, x + 1)
```
This tells us that we have generators that are both linear in x so the original must be a polynomial in x, too.

**Comment by sohaa-khan11** [comment]
poly(expand=True) sounds safe for is_polynomial. and also useful for degree, as long as cancellation forces a fallback. fastpath+fallback approach seems good.

**Comment by sohaa-khan11** [comment]
There was a merge conflict after recent upstream changes, which have been resolved. Some tests are failing due to those changes. I can address the failing tests. Please let me know if you’d prefer a different approach or if there’s anything specific you’d like me to focus on next. 
I would love to help you guys out :)

**Comment by smichr** [comment]
I am trying to wrap something around your `_fast_degree`. Any changes I make here can always be removed by you and your work is not lost. Hopefully this back and forth doesn't make you nervous. We can clean it up at the end. But I think it's going to be robust in the end.

**Comment by smichr** [comment]
@sohaa-khan11 it looks like you already know how to resolve conflicts; do you know how to merge commits together? If so, you might put the last 4 that I did into a single one. I know tests won't pass but it is a good exercise.

File: sympy/polys/tests/test_polytools.py [hunk 5]
```diff
@@ -1350,7 +1350,7 @@ def test_Poly_degree():
 
     # optimization
     big = 2  # make this number 1000 when full expansion can be avoided
-
+
```
Comment (smichr) [inline sympy/polys/tests/test_polytools.py]:
```suggestion

```

**Comment by sohaa-khan11** [comment]
Yes! i do know how to resolve conflicts and merge commits. i have squashed the 4 commits you did into a single WIP commit. Tests are still failing, I'll work on fixing them next.
Thanks for the guidance :)

**Comment by sohaa-khan11** [comment]
I have fixed the ruff failure. One test is still failing — I’m trying to solve it. 
If anything specific, you can guide me, or I will continue fixing it...

**Comment by sohaa-khan11** [comment]
I checked the performance for the original issue (#6322).
degree((x + 1)**10000, gen=x) now runs in about 0.08s on my machine, so it’s clearly hitting the fast path and no longer expanding the polynomial.
Also confirmed that degree(I) now raises TypeError, and test_Poly_degree is passing locally.

**Comment by jhaayushkumar** [comment]
something wrong with your git commit see the file changes..

**Comment by smichr** [comment]
I will start looking this over again. I didn't forget about this.
